### ISSUE TITLE
ModelCheckpoint misformats filename if keys have shared substrings

### ISSUE THREAD

**Issue by sidnarayanan** [Lightning-AI/lightning#16290]
### Bug description

Using a filename such as this causes `ModelCheckpoint._format_checkpoint_name()` to replace shared substrings with `0[shared_substr]` in all keys with the substring.
```
    filename = "{epoch:04d}_{step:010d}_{epoch_test:.3g}"
```

In this case, the filename gets mangled to the following before metric values are inserted:
```
{0[epoch]:04d}_{0[step]:010d}_{0[epoch]_test:.3g}
```

The bug happens with and without `auto_insert_metric_name`.

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

I have a pretty simple patch. Happy to open a PR if this makes sense:

```python
    @classmethod
    def _format_checkpoint_name(
        cls,
        filename: Optional[str],
        metrics: Dict[str, Tensor],
        prefix: str = "",
        auto_insert_metric_name: bool = True,
    ) -> str:
        if not filename:
            # filename is not set, use default name
            filename = "{epoch}" + cls.CHECKPOINT_JOIN_CHAR + "{step}"

        # check and parse user passed keys in the string
        groups = re.findall(r"{([^\:}]*)(:?[^}]*)}", filename)
        if len(groups) >= 0:
            for name, fmt in groups:
                if auto_insert_metric_name:
                    # {name:fmt} to name={name:fmt}
                    metric_prefix = name + "="
                else:
                    metric_prefix = ""

                # 0[name] <- support for dots: https://stackoverflow.com/a/7934969
                old_substr = "{" + name + fmt + "}"
                new_substr = metric_prefix + "{0[" + name + "]" + fmt + "}"
                filename = filename.replace(old_substr, new_substr)

                if name not in metrics:
                    metrics[name] = torch.tensor(0)
            filename = filename.format(metrics)

        if prefix:
            filename = cls.CHECKPOINT_JOIN_CHAR.join([prefix, filename])

        return filename

```

cc @carmocca @awaelchli

**Comment by awaelchli** [comment]
@sidnarayanan By shared substring you mean simply the fact that in this example, `filename = "{epoch:04d}_{step:010d}_{epoch_test:.3g}"` the two metric names "epoch" and "epoch_test" both share the same "epoch" as substring. Is that what is causing an issue?

Thanks for looking into a patch. Feel free to validate it against our tests, by running

```py
py.test -v tests/tests_pytorch/checkpointing
```

A PR would be welcome.

**Comment by bkiat1123** [comment]
@awaelchli I would like to take this.

**Comment by awaelchli** [comment]
@bkiat1123 Feel free to work on this. You could try @sidnarayanan's patch that was suggested above and check whether it passes the existing tests.
